### ISSUE TITLE
git brainstorming

### ISSUE THREAD

**Issue by bwalsh** [ACED-IDP/gen3_util#40]
### From 2023-01-03 discussion

<img width="952" alt="image" src="https://github.com/ACED-IDP/gen3_util/assets/47808/515a18fc-c8d8-4cd2-a934-3c984ae1d8c5">

#### init

> As an ACED user, when I want to start a new project, I need a simple way to create project structure.

[x] done

gen3_util `init`

```
Usage: gen3_util init [OPTIONS]

  Create project, both locally and on remote.

Options:
  --project_id TEXT  Gen3 program-project
  --help             Show this message and exit.

```

* create common directory structure [DATA/, META/, .g3t/]
* localize config file and state directory in hidden dir 
* issue requestor commands to create project in remote, (ready for signing)

#### clone

> As an ACED user, when I want to work with a project locally, I need a simple way to retrieve meta data and store configuration in well known locations.

[x] done

gen3_util `clone`

```
Usage: gen3_util clone [OPTIONS]

  Clone meta and files from remote.

Options:
  --project_id TEXT             Gen3 program-project
  --data_type [meta|files|all]  Clone meta and/or files from remote.
                                [default: all]
```

* creates directories, the same as `init`
* downloads meta, files 

#### commit

> As an ACED user, when I think I'm complete with  my contributions, I need a single command to record a comment and run validation and sanity checks.

gen3_util `commit`

#### status

> As an ACED user, when I'm making changes, I need a quick way to summarize my changes versus the remote project

gen3_util `status`

#### log

> As an ACED user, when I need to see the history of a project, I need a quick way to summarize the contributions

gen3_util `log`

#### push

> As an ACED user, when I'm ready to publish my contributions, I need a single command to upload files and meta data.

gen3_util `push`

> As an ACED user, I may be using gen3_utils on a system with the data already on the file system - or I may be working on a system that will retrieve the data.  It would be useful if the project structure could incorporate symlinks

Possible implementation points: `file add`, `clone`, `init`, `pull`

**Comment by bwalsh** [comment]
Re. `commit`

> Q:  Where should we direct the user to store `contributions`?  In other words,  `DATA/...` has all files, `META/...` has ndjson of existing meta, **where should new meta and files  be stored?**  

**tl;dr** - where should new metadata contributions be stored?  
a) in a pre-determined location, 
b) any place the user wants
Thoughts?


#### Option 1:  pre-determined location 

When:

* `META-STAGING/`  has new ndjson
* `DATA/` has new files

Then:

* `commit -m ...` will:

  * validate all metadata in META-STAGING/,
  * write commit message to new Task record.

* `push` will:
   *  upload metadata from META-STAGING
   * new files in DATA/, 


#### Option 2:  Any location 

When:

* `<my-directory>/`  has new ndjson
* `DATA/` has new files

Then:

* `commit <my-directory>/ -m ...` will:

  * validate all metadata in `<my-directory>/`,
  * write commit message to new Task record.

* `push` will:
   *  upload metadata from `<my-directory>/`
   * new files in DATA/, 


#### After `push`

* `pull` will:
  *  download all metadata to META/
  * download all new files to DATA/

**Comment by lbeckman314** [comment]
> tl;dr - where should new metadata contributions be stored?
> a) in a pre-determined location,
> b) any place the user wants
> Thoughts?

I like Option 1 in that it gives the user a structured expectation of where the metadata will exist (less uncertainty perhaps), but Option 2 is more flexible. `META-STAGING/` could even be an env var that the user could be able set, with `<project root>/META-STAGING` being the default:

```sh
# default
g3t commit

# commit flag
g3t commit --meta /tmp/example-project

# env var
export META-STAGING=/tmp/example-project
g3t commit
```
